app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshake.kt

```diff
@@ -0,0 +1,151 @@
+ * initiates the file transfer through [SenderDevice] on the sender and using
+ * [ReceiverDevice] on the receiver.
+ */
+internal class PeerGroupHandshake(private val wifiDirectManager: WifiDirectManager) {
+  private val HANDSHAKE_MESSAGE = "Request Kiwix File Sharing"
+  suspend fun handshake(): InetAddress? = withContext(Dispatchers.IO) {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Handshake in progress")
+    }
+    if (wifiDirectManager.isGroupFormed && wifiDirectManager.isGroupOwner && this.isActive) {
+      try {
+        ServerSocket(PEER_HANDSHAKE_PORT)
+          .use { serverSocket ->
+            serverSocket.reuseAddress = true
+            val server = serverSocket.accept()
+            val objectInputStream = ObjectInputStream(server.getInputStream())
+            val kiwixHandShakeMessage = objectInputStream.readObject()
+
+            // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+            return@withContext if (isKiwixHandshake(kiwixHandShakeMessage) && this.isActive) {
+              if (BuildConfig.DEBUG) {
+                Log.d(TAG, "Client IP address: " + server.inetAddress)
+              }
+              exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream())
+              server.inetAddress
+            } else {
+              // Selected device is not accepting wifi direct connections through the kiwix app
+              null
+            }
+          }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return@withContext null
+      }
+    } else if (wifiDirectManager.isGroupFormed && this.isActive) { // && !groupInfo.isGroupOwner
+      try {
+        Socket().use { client ->
+          client.reuseAddress = true
+          client.connect(
+            InetSocketAddress(
+              wifiDirectManager.groupOwnerAddress.hostAddress,
+              PEER_HANDSHAKE_PORT
+            ), 15000
+          )
+          val objectOutputStream = ObjectOutputStream(client.getOutputStream())
+          // Send message for the peer device to verify
+          objectOutputStream.writeObject(HANDSHAKE_MESSAGE)
+          exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream())
+          return@withContext wifiDirectManager.groupOwnerAddress
+        }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return@withContext null
+      }
+    }
+    return@withContext null
+  }
+
+  private fun isKiwixHandshake(handshakeMessage: Any): Boolean =
+    HANDSHAKE_MESSAGE == handshakeMessage
+
+  private fun exchangeFileTransferMetadata(outputStream: OutputStream, inputStream: InputStream) {
+    if (wifiDirectManager.isFileSender) {
+      try {
+        ObjectOutputStream(outputStream).use { objectOutputStream ->
+          // Send total number of files which will be transferred
+          objectOutputStream.writeObject("" + wifiDirectManager.totalFilesForTransfer)
```

Comment:
can use string interpolation... also why are we not just writing an int and taking an int out on the other side?

Reply:
done

Reply:
an objectOutputStream unsurprisingly lets you write objects, arbitrarily. An objectInputStream lets you read objects but you don't inherently know what type that object is.
we write a string here
```
 objectOutputStream.writeObject("" + wifiDirectManager.totalFilesForTransfer)
```
and then we read an object, call toString and then parse that string as an int
```
 val totalFilesObject = objectInputStream.readObject().toString()
        if (totalFilesObject.javaClass == String::class.java) {
          val total: Int = totalFilesObject.toInt()
```
which is quite a bit of nonsense because we could just
```
objectOutputStream.writeObject(wifiDirectManager.totalFilesForTransfer)
...
 (objectInputStream.readObject() as? Int).let{ totalFilesForTransfer-> }
```

`totalFilesObject.javaClass == String::class.java` this line in particular is also a nonsense because we have just called toString, a method available on all objects, of course it is a string

Reply:
done [here](https://github.com/kiwix/kiwix-android/pull/2389/commits/3f554492663bc2ed22695bfd4e894925a38cfc26)

Reply:
it is not done, you are still writing a string here

Reply:
nvm, stupid github showing old diffs